contrib/platform/test/com/sun/jna/platform/win32/MsiTest.java

```diff
@@ -33,52 +33,62 @@ public static void main(String[] args) {
+        assertCallSucceededMsiError("MsiGetProductCode", Msi.INSTANCE.MsiGetProductCode(component, productBuffer));
 
         String product = new String(productBuffer).trim();
-        assertTrue(product.length() > 0);
+        assertFalse("Product is empty", product.isEmpty());
     }
 
     public void testMsiLocateComponentW() {
         char[] componentBuffer = new char[40];
-        assertTrue(W32Errors.ERROR_SUCCESS == Msi.INSTANCE.MsiEnumComponents(new WinDef.DWORD(0), componentBuffer));
+        assertCallSucceededMsiError("MsiEnumComponents", Msi.INSTANCE.MsiEnumComponents(new WinDef.DWORD(0), componentBuffer));
         String component = new String(componentBuffer).trim();
+        assertFalse("Component is empty", component.isEmpty());
 
         char[] pathBuffer = new char[WinDef.MAX_PATH];
         IntByReference pathBufferSize = new IntByReference(pathBuffer.length);
-        Msi.INSTANCE.MsiLocateComponent(component, pathBuffer, pathBufferSize);
+        int installState = Msi.INSTANCE.MsiLocateComponent(component, pathBuffer, pathBufferSize);
 
         String path = new String(pathBuffer, 0, pathBufferSize.getValue()).trim();
-        assertTrue(path.length() > 0);
+        assertFalse("Path is empty. MsiLocateComponent returned " + MsiUtil.formatMessage(installState), path.isEmpty());
     }
 
     public void testMsiGetComponentPathW() {
         char[] componentBuffer = new char[40];
-        assertTrue(W32Errors.ERROR_SUCCESS == Msi.INSTANCE.MsiEnumComponents(new WinDef.DWORD(0), componentBuffer));
+        assertCallSucceededMsiError("MsiEnumComponents", Msi.INSTANCE.MsiEnumComponents(new WinDef.DWORD(0), componentBuffer));
         String component = new String(componentBuffer).trim();
+        assertFalse("Component is empty", component.isEmpty());
 
         char[] productBuffer = new char[40];
-        assertTrue(W32Errors.ERROR_SUCCESS == Msi.INSTANCE.MsiGetProductCode(component, productBuffer));
+        assertCallSucceededMsiError("MsiGetProductCode", Msi.INSTANCE.MsiGetProductCode(component, productBuffer));
 
         String product = new String(productBuffer).trim();
-        assertTrue(product.length() > 0);
+        assertFalse("Product is empty", product.isEmpty());
 
         char[] pathBuffer = new char[WinDef.MAX_PATH];
         IntByReference pathBufferSize = new IntByReference(pathBuffer.length);
-        Msi.INSTANCE.MsiGetComponentPath(product, component, pathBuffer, pathBufferSize);
+        int installState = Msi.INSTANCE.MsiGetComponentPath(product, component, pathBuffer, pathBufferSize);
 
         String path = new String(pathBuffer, 0, pathBufferSize.getValue()).trim();
-        assertTrue(path.length() > 0);
+        assertFalse("Path is empty. MsiGetComponentPath returned " + MsiUtil.formatMessage(installState), path.isEmpty());
+    }
+
+    private void assertCallSucceededMsiError(String message, int state) {
```

Comment:
Can you tell me how this differs from just using this in the previous calls?
```
assertEquals(String.format("TestNameHere failed with: %s",  MsiUtil.formatMessage(state),
    W32Errors.ERROR_SUCCESS, state);
```

Reply:
Let's use MsiEnumComponents as an example.
As long as MsiEnumComponents is only called once and its return value is used for both checking for success code and forwarded to the formatting routine, there is no difference.

The reason why I did it like this is that I used AbstractWin32TestSupport.assertCallSucceeded() as an template and it looked similar.

Reply:
You could have just called `AbstractWin32TestSupport` :).   Except you're adding the error message.   Not sure I particularly like using a "Util" for that, will have to think about it.

Reply:
The reason that I put it in the util class was that it could be useful to end-user applications too, but right now, it could be moved to the test class in order to still have a nicer error message.
Well, if I were to call AbstractWin32TestSupport.assertCallSucceeded(), then I would use GetLastError() that is wrong here... ;-)

Reply:
True point. I'm [thinking of the scalability of case statements for Error messages](https://github.com/java-native-access/jna/blob/d471b15353a9dbcc5a29064385e4e50ca5d257c5/contrib/platform/src/com/sun/jna/platform/linux/ErrNo.java).  I'm not sure as a library we need to include the full description, when the enum name is probably descriptive enough.

Reply:
Maybe I'm missing something, but this is just an ordinary integer, not an enum?
Although the complete list of possible return values are unknown, MSDN just lists a few codes, so I don't think the list will be that huge. In any case, your comment made me take a second look at the list and it appears that the formatMessage() function is missing some values.
I'll add them and push again.

Should I also move the function to the test class to not create this mapping for the end-user application?

Reply:
> MSDN just lists a few codes, so I don't think the list will be that huge.

This is less a point for this specific PR as it is for error codes in general, and an overall philosophy of what a library should provide vs. what a user-facing application should provide.  

There are 8 error messages in this PR, but there are 131 Linux error codes, and I haven't even counted how many error codes are in the 40K+ lines of the `WinError` class. A few thousand at least. A switch/case statement won't work for that.  

It's entirely reasonable, and even desirable, for a library to expose an error code to applications using it.  But translating codes to text is not necessarily the responsibility of a library.   I don't think JNA does this generally, so if we choose to do so now, we set a precedent.  While a switch with 8 cases doesn't look too bad in this case, the header file for Linux error codes that I linked above would, if we copied the same format/implementation, require 131 "case" options on the switch, expanding the library size with 131 strings that (hopefully) nobody will ever see.  Windows errors total something in the 3000 (plus or minus) order of magnitude.

Having a more detailed error description for a unit test is slightly more reasonable, but I think it's a waste of time. In the best (or even "usual") case scenario, all tests will pass and no one will ever see the error message.  The question is whether a developer writing code that causes the test to fail will have enough information to figure out their mistake and fix it.  In this case, simply identifying the error code (decimal or hex or both) is probably sufficient.  Unlike an end user, the developer is already working directly with the methods involved and is more likely to know where to look up the error code.

You're correct that it's a constant, not an enum, so we can't really extract the name.  I still think the error code itself is sufficient for the (hopefully rare) case where an error occurs.

Reply:
Okay. I get your point.

I'll update the PR.

Reply:
Looks good, ready to merge.  But see comment below. 

I was looking into more generic error message functionality, though, per the earlier version of the PR.  It looks like, on Windows, the [`FormatMessage`](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-formatmessage) function is the proper way to look up/display textual descriptions of error codes.  This is already implemented in [`Kernel32Util#formatMessage`](https://github.com/java-native-access/jna/blob/8a44e4c290aeda3a19d6cdad888cd5ccbe4375b9/contrib/platform/src/com/sun/jna/platform/win32/Kernel32Util.java#L188) for the WinError error codes (resolving my earlier concern).

It looks like you could do something similar, fetching the Msi Module using [GetModuleHandleEx](https://docs.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulehandleexw).

Reply:
FWIW in a past life of Windows development at MSFT we used to call `FormatMessage` to show the error to the developer in unit tests to avoid displaying "expected S_OK, but got 0xBAADFOOD` in a macro style (C++, `ChkHR(DoSomethingThatReturnsHR, "I failed to Do Something")`; . In case of failure it would say "I failed to Do Something, the spline could not be reticulated".

So +1 to @dbwiddis suggestion.

Reply:
I tried to do something with GetModuleHandleEx, but it did not work.
Maybe that part could be different PR and this PR could be merged as-is?